Attempt to add py2exe

HeavyTony2 opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team][1] using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel][2] if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [ ] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [ ] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [ ] Source is from official source.
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [ ] Build number is 0.
- [ ] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes```) and found some lint.

Here's what I've got...


For **recipes**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/py2exe```) and found some lint.

Here's what I've got...


For **recipes/py2exe**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/py2exe```) and found some lint.

Here's what I've got...


For **recipes/py2exe**:

 * Non noarch packages should have python requirement without any version constraints.
 * Non noarch packages should have python requirement without any version constraints.

For **recipes/py2exe**:

 * License is not an SPDX identifier (or a custom LicenseRef) nor an SPDX license expression.

Documentation on acceptable licenses can be found [here]( https://conda-forge.org/docs/maintainer/adding_pkgs.html#spdx-identifiers-and-expressions ).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/py2exe```) and found some lint.

Here's what I've got...


For **recipes/py2exe**:

 * There are 2 too many lines.  There should be one empty line at the end of the file.

For **recipes/py2exe**:

 * License is not an SPDX identifier (or a custom LicenseRef) nor an SPDX license expression.

Documentation on acceptable licenses can be found [here]( https://conda-forge.org/docs/maintainer/adding_pkgs.html#spdx-identifiers-and-expressions ).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/py2exe```) and found it was in an excellent condition.

HeavyTony2 commented:
I removed Future as specific dependancies and replaced the install script by the normal one. I tested with a workflow on a private repo and it did work. The problem was on my pc I guess.

HeavyTony2 commented:
For the recap, that one worked https://github.com/conda-forge/staged-recipes/pull/17596/commits/ead9813e2ff236eb938a18ff1afed75fba1a756d
That one didn't https://github.com/conda-forge/staged-recipes/pull/17596/commits/d8f2e50dcc7ebab13b4d949e8a606d6e21675b06

I'm now trying that one
https://github.com/conda-forge/staged-recipes/pull/17596/commits/7045956a21d99965d53486b75d4344b0ba548c3f

HeavyTony2 commented:
ping @conda-forge/staged-recipes

HeavyTony2 commented:
> @ocepaf The current errors are caused by the way the compiler parses a source file. Do you know a way to fix it? https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=446887&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=06d62fd2-ab49-5add-9ee5-83b6882d7d2b&l=2921

Whoops, I mistyped your username @ocefpaf 

I also noticed something else:  This is from Azure logs.
 `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\cl.exe" /c /nologo **/O2** /W3 /GL /DNDEBUG /MD -DPYTHONDLL=\\\"python37.dll\\\" -D_CRT_SECURE_NO_WARNINGS=1 -DPy_BUILD_CORE=1 -IC:\bld\py2exe_1643136732493\_h_env\include -IC:\bld\py2exe_1643136732493\_h_env\Include -IC:\bld\py2exe_1643136732493\_h_env\Library\include "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\winrt" /d1trimfile:C:\bld\py2exe_1643136732493\work /Tcsource/start.c /Fobuild\temp.win-amd64-3.7\Release\source/start.obj "-IC:\Program Files\Microsoft SDKs\Windows\v7.0\Include" "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include" "-IC:\Program Files (x86)\Windows Kits\10\Include\10.0.10586.0\ucrt`
 
 Now, a log from a workflow on a private repo of mine
  `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\cl.exe /c /nologo **/Ox** /W3 /GL /DNDEBUG /MD -DPYTHONDLL="python39.dll" -D_CRT_SECURE_NO_WARNINGS=1 -DPy_BUILD_CORE=1 -IC:\Miniconda\conda-bld\py2exe_1643150042326\_h_env\include -IC:\Miniconda\conda-bld\py2exe_1643150042326\_h_env\include -IC:\Miniconda\conda-bld\py2exe_1643150042326\_h_env\Library\include -IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE -IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE -IC:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.22000.0\winrt /d1trimfile:C:\Miniconda\conda-bld\py2exe_1643150042326\work /Tcsource/start.c /Fobuild\temp.win-amd64-3.9\Release\source/start.obj -IC:\Program Files\Microsoft SDKs\Windows\v7.0\Include -IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include -IC:\Program Files (x86)\Windows Kits\10\Include\10.0.10586.0\ucrt`  
  
 The optimization flag isn't the same. Could it be related and is it overridable in a batch script?

jakirkham commented:
Looks similar to issue ( https://github.com/py2exe/py2exe/issues/110 ). The author notes this may be fixed in the more recent [`0.11.0.1` release]( https://pypi.org/project/py2exe/0.11.0.1/ ). Perhaps try upgrading to see if that fixes it

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/py2exe```) and found some lint.

Here's what I've got...


For **recipes/py2exe**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/py2exe```) and found it was in an excellent condition.

HeavyTony2 commented:
> Looks similar to issue ( [py2exe/py2exe#110](https://github.com/py2exe/py2exe/issues/110) ). The author notes this may be fixed in the more recent [`0.11.0.1` release](https://pypi.org/project/py2exe/0.11.0.1/). Perhaps try upgrading to see if that fixes it

Makes sense. I was working with an a older version because the newer ones didn't have the py2exe_distutils.py file. In fact, they still had it, but only on Github, not on Pypi. So, I ultimately switched to Github repos, but forgot to upgrade.

jakirkham commented on recipes/py2exe/meta.yaml:
```diff
@@ -0,0 +1,52 @@
+  skip: True  # [not win]
+  script: {{ PYTHON }} -m pip install . -vv
+
+requirements:
```
Also we need this to setup the C compiler

```suggestion
requirements:
  build:
    - {{ compiler('c') }}
```

ocefpaf commented on recipes/py2exe/meta.yaml:
```diff
@@ -0,0 +1,56 @@
+  entry_points:
+    - build_exe = py2exe.build_exe:main
+  skip: True  # [ not win ]
+  skip: True  # [ py < 38 ]
```
setup.py says `>=3.6, <3.11`.

HeavyTony2 replied:
Yup, but I noticed this:

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=447416&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=06d62fd2-ab49-5add-9ee5-83b6882d7d2b&l=1750

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=447416&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=06d62fd2-ab49-5add-9ee5-83b6882d7d2b&l=3248

HeavyTony2 commented:
I'm trying with 3.9 only. This is for testing purpose.

HeavyTony2 commented:
Found the problem for 3.7 and 3.8....Conda-Forge :/

jakirkham commented:
Are we sure about that? Most of the packages here are `noarch` so don't vary between Python versions.

Wondering if it is something else like how `setuptools` and `distutils` have been evolving. What if we pin to an older `setuptools` version like `setuptools <60`?

HeavyTony2 commented:
> Are we sure about that? Most of the packages here are `noarch` so don't vary between Python versions.
> 
> Wondering if it is something else like how `setuptools` and `distutils` have been evolving. What if we pin to an older `setuptools` version like `setuptools <60`?

That sounds like a good idea. 
1- My block conda-forge with the conda_build_config.yaml file doesn't work anywa.
2- I just gave a a try on conda-incubator/setup-miniconda@v2 with conda-forge as channel and setuptools 58.0.4 as host dependancies and it did work.

HeavyTony2 commented:
Thanks a lot @jakirkham It dit work!

jakirkham commented on recipes/py2exe/meta.yaml:
```diff
@@ -0,0 +1,59 @@
+  build:
+    - {{ compiler('c') }}
+  host:
+    - setuptools <60
```
Probably worth adding a comment here now that we know this is what is needed. Would also suggest raising an upstream issue about this and linking it here

HeavyTony2 replied:
So I basically open an issue to explain the situation with newer setuptools version and also about cachetools and pefile?

HeavyTony2 commented:
https://github.com/py2exe/py2exe/issues/124

jakirkham commented on recipes/py2exe/meta.yaml:
```diff
@@ -0,0 +1,59 @@
+  build:
+    - {{ compiler('c') }}
+  host:
+    - setuptools <60
+     # Newer versions cause parsing issues for PY37&38
```
Would suggest this slight tweak. Otherwise looks good :)

```suggestion
     # Newer setuptools versions cause parsing issues for PY37&38
     # xref: https://github.com/py2exe/py2exe/issues/124
    - setuptools <60
```

HeavyTony2 commented:
@jakirkham Can I add you to the maintainers list?

jakirkham commented:
Thanks for all the hard work here @HeavyTony2! Also thanks for the help reviewing Filipe 😄

Think I'll hold off on joining the maintainer list simply because I don't have access to a Windows machine, which makes it hard to debug any issues that would arise with a Windows specific package. That said, happy to watch the feedstock and have gone ahead and subscribed to notifications there 🙂
